protocol/transaction_validator.go

```diff
@@ -0,0 +1,317 @@
+package protocol
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	coretypes "github.com/status-im/status-go/eth-node/core/types"
+	"github.com/status-im/status-go/eth-node/crypto"
+	"github.com/status-im/status-go/eth-node/types"
+	"go.uber.org/zap"
+	"math/big"
+	"strings"
+)
+
+const (
+	transferFunction        = "a9059cbb"
+	tokenTransferDataLength = 68
+	transactionHashLength   = 66
+)
+
+type TransactionValidator struct {
+	persistence *sqlitePersistence
+	addresses   map[string]bool
+	logger      *zap.Logger
+}
+
+type TransactionToValidate struct {
+	TransactionHash string
+	CommandID       string
+	MessageID       string
+	RetryCount      int
+	// First seen indicates the whisper timestamp of the first time we seen this
+	FirstSeen uint64
+	// Validate indicates whether we should be validating this transaction
+	Validate  bool
+	Signature []byte
+	From      *ecdsa.PublicKey
+}
+
+func NewTransactionValidator(addresses []types.Address, persistence *sqlitePersistence, logger *zap.Logger) *TransactionValidator {
+	addressesMap := make(map[string]bool)
+	for _, a := range addresses {
+		addressesMap[strings.ToLower(a.Hex())] = true
+	}
+	logger.Debug("Checking addresses", zap.Any("addrse", addressesMap))
+
+	return &TransactionValidator{
+		persistence: persistence,
+		addresses:   addressesMap,
+		logger:      logger,
+	}
+}
+
+type EthClient interface {
+	TransactionByHash(context.Context, types.Hash) (coretypes.Message, bool, error)
+}
+
+func (t *TransactionValidator) verifyTransactionSignature(from *ecdsa.PublicKey, address types.Address, transactionHash string, signature []byte) error {
+	publicKeyBytes := crypto.FromECDSAPub(from)
+
+	if len(transactionHash) != transactionHashLength {
+		return errors.New("wrong transaction hash lenght")
+	}
+
+	hashBytes, err := hex.DecodeString(transactionHash[2:])
+	if err != nil {
+		return err
+	}
+	signatureMaterial := append(publicKeyBytes, hashBytes...)
+
+	// We take a copy as EcRecover modifies the byte slice
+	signatureCopy := make([]byte, len(signature))
+	copy(signatureCopy, signature)
+	extractedAddress, err := crypto.EcRecover(context.Background(), signatureMaterial, signatureCopy)
+	if err != nil {
+		return err
+	}
+
+	if extractedAddress != address {
+		return errors.New("failed to verify signature")
+	}
+	return nil
+}
+
+func (t *TransactionValidator) validateTokenTransfer(parameters *CommandParameters, transaction coretypes.Message) (*VerifyTransactionResponse, error) {
+
+	data := transaction.Data()
+	if len(data) != tokenTransferDataLength {
+		return nil, errors.New(fmt.Sprintf("wrong data length: %d", len(data)))
+	}
+
+	functionCalled := hex.EncodeToString(data[:4])
+
+	if functionCalled != transferFunction {
+		return &VerifyTransactionResponse{}, nil
```

Comment:
This line looks suspicious. Is it correct?

Reply:
Yes, it returns `Valid: false` which means the transaction is not valid and it should not be validated anymore

Reply:
I see. So maybe at the top we can define something like this:
```go
invalidResp = &VerifyTransactionResponse{Valid: false}
// ...
return invalidResp, nil
```

Reply:
cool, will do that, thanks.
For now we won't provide any details (not used by the client), but we might do eventually